### PR TITLE
Initial plumbing of backward autodiff in the frontend.

### PR DESCRIPTION
**Pull request by behindthepixels**
This is a WIP PR of backward autodiff implementation.

### CODE REVIEW

File: source/slang/slang-ir-diff-jvp.cpp [hunk 1]
```diff
@@ -2518,4 +2518,1954 @@ void stripAutoDiffDecorations(IRModule* module)
 }
 
 
+
```
Comment (csyonghe) [inline source/slang/slang-ir-diff-jvp.cpp:2521]: We probably want to move this to its own file.

**Review by csyonghe — COMMENTED** [review]
The plumbing changes looks good! We probably want to move the backward pass into its own file and pull things that can be shared between the two passes into a separate `slang-ir-auto-diff-shared.h/cpp`.

**Review by csyonghe — APPROVED** [review]

File: source/slang/slang-ir-diff-jvp.cpp (hunk 1 above)
Comment (behindthepixels) [reply source/slang/slang-ir-diff-jvp.cpp:2521]: Yeah I can do some refactor after getting some basic functionality working.

**Comment by csyonghe** [comment]
Superceded by PR #2514.

**Pull request closed without merging**